### ISSUE TITLE
Feature Request: Message Queue / Event Bus

### ISSUE THREAD

**Issue by phpnode** [orientechnologies/orientdb#2652]
I've discussed this separately with Luca & Artem in the past and both agreed this was a cool idea, so here's an issue:

It would be relatively straightforward to add Event / Messaging capabilities to OrientDB. More and more (web) applications need them these days, especially with "realtime" apps becoming more prevalent. At the moment most developers either use redis's pub/sub channels, or something like RabbitMQ when durability / deliverability guarantees are required.

Redis is super easy to use in this way but most logic must remain in the application and it's hard to make it truly reliable.

RabbitMQ is relatively complicated and awkward to set up and integrate correctly, it's also Yet Another Thing To Learn.

I propose that we introduce the concept of "Transient Edges" into OrientDB. A transient edge is one that is delivered to subscribers rather than being persisted in the database. This can be used to represent messages, events etc. Transient Edges can have a TTL, with edges expiring after a certain time. An edge class can be made transient by using the following:

> Note: All syntax here is of course up for debate. Also for convenience I'm using `LET` and variables in the examples, whereas in real life this would not usually be practical. Wherever I'm using a variable it should be possible to use a `SELECT` query or `RID` instead.

```sql
CREATE CLASS MyThing EXTENDS E;
ALTER CLASS MyThing TRANSIENT true; -- calling this on a non edge class should throw an error
```

A class wide TTL can be specified by using:

```sql
ALTER CLASS MyThing TTL 600; -- 10 minutes
```

## Examples

For the purposes of demonstration, let's imagine some collaborative workflow tool like github which has Organisations which consist of many Teams. Users belong to one Organisation and zero or more Teams within that Organisation.

**Setup**

```sql
CREATE CLASS Organisation EXTENDS V;
CREATE CLASS Team EXTENDS V;
CREATE CLASS Member EXTENDS E;

LET admin = SELECT FROM #5:0;
LET reader = SELECT FROM #5:1;

LET org = CREATE VERTEX Organisation SET name = "orientdb"; // #20:0
LET developers = CREATE VERTEX Team SET name = "developers", org = $org; // #21:0

-- Assign users to the org and team.
CREATE EDGE Member FROM $admin TO $org;
CREATE EDGE Member FROM $reader TO $org;
CREATE EDGE Member FROM $admin TO $developers;
CREATE EDGE Member FROM $reader TO $developers;
```

### Example 1. Instant Messaging.
As a user I want to be able to send instant messages to other users, teams or entire organisations.

```sql
-- A custom edge class, used to represent instant messages.
CREATE CLASS InstantMessage EXTENDS E;

-- Put the edge class into transient mode.
ALTER CLASS InstantMessage TRANSIENT true;

-- Expire messages after 10 minutes by default
ALTER CLASS InstantMessage TTL 600;


-- Create a subscription for InstantMessages for the reader.
SUBSCRIBE InstantMessage FROM $org, $developers, $reader; -- returns a temporary RID for the subscription


-- Send a message from admin to reader
PUBLISH InstantMessage TO $reader FROM $admin CONTENT { "message": "Hello there", "isPrivate": true };

-- Send a message from reader to the developers team
PUBLISH InstantMessage TO $developers FROM $reader SET message = "Hi Everybody!";

-- Send a message from reader to the org, with a TTL
PUBLISH InstantMessage TO $org FROM $reader SET message = "this message will self destruct in 6 seconds" TTL 6;

```

It should also be possible to *listen* for these edges and bind to events, for example:

```sql
-- Add Custom Event Listeners

ON InstantMessage Organisation MyFunc(isPrivate, message, @out);
ON InstantMessage Organisation INSERT INTO MessageLog fields(doc()), timestamp = now();

-- Remove Custom Event Listeners
OFF InstantMessage Organisation MyFunc(isPrivate, message, @out);
OFF InstantMessage Organisation INSERT INTO MessageLog fields(doc()), timestamp = now();
```

### Example 2. Background Workers / Tasks

When certain kinds of activity happen in the application, we want to send notification emails. We use background workers to send these emails. At any time, there may be zero or more workers online, email notifications *must* be sent once and only once.

```sql
CREATE CLASS EmailTask EXTENDS E;
ALTER CLASS EmailTask TRANSIENT true;
ALTER CLASS EmailTask MINRECIPIENTS 1;
ALTER CLASS EmailTask MAXRECIPIENTS 1;
ALTER CLASS EmailTask ACK true; -- clients must "ack" messages of this type

-- Create a class to represent the various services in the app
CREATE CLASS Service EXTENDS V;

-- Create the email service
LET emailService = INSERT INTO Service SET name = "emailService";

-- When an organisation is created, send a welcome email.
ON CREATE Organisation PUBLISH EmailTask FROM @rid TO $emailService SET emailType = "welcome", recipient = owner;

-- When a user joins a team, email the team.
ON CREATE EDGE Team WHERE @class = Member PUBLISH EmailTask FROM @in TO $emailService SET emailType = "newTeamMember", recipients = @in.in_Member, team = @in;

-- Workers then subscribe to the service, listening for any kind of event.
LET $subscription = SUBSCRIBE * FROM $emailService; -- return a temporary RID;

-- They can later unsubscribe using
UNSUBSCRIBE $subscription;
```

### Why reuse edges instead of adding a new custom event type.
Because it's easier and we get a lot of things like validation etc for free.

### What happens if I `SELECT * FROM SomeTransientEdgeClass`? 
This returns all the edges which have not yet been delivered or have not yet expired.

### How would clients implement this?
The binary protocol already supports the concept of `PUSH_DATA`, this can be expanded to support messages / events too. We'd also need to create a new operation for acknowledging messages which require it. This op would just take a `sessionId` and a `RID`.

**Comment by saeedtabrizi** [comment]
Good Idea and many useful for me :) :+1:  
nowadays , online apps needs to be connected and some instant  messaging functions . 
please change milestone to closest release .

**Comment by mattaylor** [comment]
Any chance of gettting this in 2.1?

**Comment by StarpTech** [comment]
That's a powerful feature to let the application work for itself without the use of external services or manuell triggers in the domain logic. +1

**Comment by lvca** [comment]
We already have an alpha version of this under the branch "livequery" (https://github.com/orientechnologies/orientdb/tree/livequery). We are deciding if include this as "experimental" in the upcoming 2.1 or postpone it to 2.2.

**Comment by lvca** [comment]
Look also at @luigidellaquila's presentation: http://www.slideshare.net/LuigiDellAquila/fully-reactive-from-data-to-ui-with-orientdb-nodejs-socketio

**Comment by mattaylor** [comment]
I vote Experimental in 2.1 

From:  Luca Garulli
Reply-To:  orientechnologies/orientdb
Date:  Monday, April 6, 2015 at 12:03 PM
To:  orientechnologies/orientdb
Cc:  Matthew Taylor
Subject:  Re: [orientdb] Feature Request: Message Queue / Event Bus (#2652)


We already have an alpha version of this under the branch "livequery" (https://github.com/orientechnologies/orientdb/tree/livequery). We are deciding if include this as "experimental" in the upcoming 2.1 or postpone it to 2.2.

**Comment by lvca** [comment]
Ok, this feature will be part of 2.1 as Experimental. In 2.2 we plan to put it as stable.

**Comment by luigidellaquila** [comment]
This feature is available as experimental in 2.1.
Moving forward to 2.2 for the final stable version

**Comment by voxadam** [comment]
Does anyone have a status update on this? I just spent far too many hours examining MQs for use in an OrientDB centric project and would truly love to see a native solution.

**Comment by lvca** [comment]
We already have Live Query in 2.1: http://orientdb.com/docs/last/Live-Query.html.

**Comment by voxadam** [comment]
@lvca, I'm sorry, somehow I managed to let the Live Query feature of 2.1 slip past me entirely. I'll blame it on the fact that I've been having entirely too much fun with all the existing features in OrientDB to keep up with all the new additions. 

In the event based worlds that I work in this will be quite handy. While I understand where @carloprad @jdillon, @malakeel, and @mattaylor are coming from I personally welcome this feature. Sure, I *could* use a dedicated queue or broker, but in my applications which tend to bend in the direction of embedded computing, I *really* appreciate not having to add yet another package to my stack if I can avoid it.

So, one last thing, in this "issue" remains to keep it from being marked as closed?

**Comment by lvca** [comment]
I'm closing this issue because live query are already in.
